microprofile/WFLY-12609_MicroProfile_JWT_1_1.adoc

```diff
@@ -0,0 +1,109 @@
+
+Additional integration may be added within https://github.com/wildfly-security/wildfly-elytron[WildFly Elytron] to bridge into the application server's security framework.
+
+=== Other Interested Projects
+
+The integration will also be making use of https://github.com/smallrye/smallrye-jwt[SmallRye JWT] as well as pulling in the https://github.com/eclipse/microprofile-jwt-auth[Eclipse MicroProfile JWT] API library.
+
+== Requirements
+
+=== Hard Requirements
+
+Primarily this enhancement will be delivered using a new subsystem `microprofile-jwt` which will be added to WildFly.  The subsystem will install a DeploymentUnitProcessor which will detect if MicroProfile JWT is required and activate a suitable security configuration as described within this proposal.
+
+At the outset activation will only occur in the following circumstances: -
+
+ * A `@org.eclipse.microprofile.auth.LoginConfig` annotation is detected with an `auth-method` of `MP-JWT`.
+ * An auth method of `MP-JWT` is specified in any web.xml.
+
+This activation scope is deliberately constrained to allow us to incrementally increase the scope at a later point. 
+
+At the point activation occurs the deployment will not make use of a configured `SecurityDomain` instead a "Virtual" `SecurityDomain` will be created to meet the requirements of the deployment and will be used across all sub-deployments.  The common `SecurityDomain` will allow for identity propagation across all sub-deployments whilst the virtual nature of the domain will make it unavailable for propagation across unrelated deployments allowing for deployment level isolation.
+
+Activation will add module dependencies as required on SmallRye JWT, Elytron JWT, and the MicroProfile JWT API.
+
+Activation will also ensure activation of the SmallRye JWT CDI extension as well as any Elytron JWT CDI extension if deemed required and any Elytron specific overrides.
+
+Finally activation will ensure the activation of the `MP-JWT` authentication mechanism for the deployment.
+
+The combination of SmallRye JWT with Elytron JWT will be the pair of projects responsible for providing the functionality described in the MicorProfile JWT specification combined with tying in the integration to the application server's security framework to ensure aspects such as identity association and establishment are correctly performed.
+
+Once an identity has been established according to the steps described by the MicroProfile JWT specification the remaining security decisions required are covered by the relevant Jakarta EE specifications in use by the various subsystems. 
+
+=== Nice-to-Have Requirements
+
+It will not be a priority of this enhancement to add support for alternative authentication mechanisms to `MP-JWT` as this brings a lot of additional complexity.  For this initial integration how we provide support for the `MP-JWT` mechanism should be largely transparent to the deployment allowing us to evolve this aspect further.
```

Comment:
For the alternative authentication mechanisms they still remain fully supported without the activation of the new MicroProfile JWT specification and at the moment the two do not come together cleanly.

As an example the motivation for using MicroProfile JWT is these tokens fully describe the resulting identity, this is why we can avoid relying on managed security configuration - the support for the alternative authentication mechanisms then triggers a dependency on managed configuration to specify things such as database and ldap identity stores etc...

Additionally once you move to mechanisms other than MP-JWT the remainder of the spec also becomes obsolete - items such as the Principal type to inject etc.. are no longer available as there was no longer an incoming token to make available / propagate.

There are a lot of options regarding alternative mechanism support but the primary focus here is adding support for MP-JWT - once this is added will open up discussions on how to better solve the alternative mechanism support - this may be enhancement to WildFly, enhancements to SmallRye JWT and hopefully even clarifications within the specification.